beacon-chain/blockchain/service.go

```diff
@@ -223,69 +164,168 @@ func (c *ChainService) CurrentActiveState() *types.ActiveState {
+	// server to stream these events to beacon clients.
+	transition := c.chain.IsCycleTransition(slot)
+	if transition {
+		c.canonicalCrystallizedStateFeed.Send(canonicalCrystallizedState)
+	}
+	c.canonicalBlockFeed.Send(canonicalBlock)
 }
 
-// run processes the changes needed every beacon chain block,
-// including epoch transition if needed.
-func (c *ChainService) run(done <-chan struct{}) {
+func (c *ChainService) blockProcessing(done <-chan struct{}) {
+	sub := c.incomingBlockFeed.Subscribe(c.incomingBlockChan)
+	defer sub.Unsubscribe()
 	for {
 		select {
-		case block := <-c.latestBeaconBlock:
-			// TODO: Apply 2.1 fork choice logic using the following.
-			vals, err := c.chain.validatorsByHeightShard()
+		case <-done:
+			log.Debug("Chain service context closed, exiting goroutine")
+			return
+		// Listen for a newly received incoming block from the sync service.
+		case block := <-c.incomingBlockChan:
+			// 3 steps:
+			// - Compute the active state for the block.
+			// - Compute the crystallized state for the block if cycle transition.
+			// - Store both states and the block into a data structure used for fork choice.
+			//
+			// Another routine will run that will continually compute
+			// the canonical block and states from this data structure using the
+			// fork choice rule.
+			var canProcess bool
+			var err error
+
+			h, err := block.Hash()
 			if err != nil {
-				log.Errorf("Unable to get validators by height and by shard: %v", err)
-				continue
+				log.Debugf("Could not hash incoming block: %v", err)
 			}
-			log.Debugf("Received %d validators by height", vals)
 
-			// Entering epoch transitions.
-			transition := c.chain.IsEpochTransition(block.SlotNumber())
-			if transition {
-				c.chain.CrystallizedState().SetStateRecalc(block.SlotNumber())
-				if err := c.chain.calculateRewardsFFG(block); err != nil {
-					log.Errorf("Error computing validator rewards and penalties %v", err)
-					continue
-				}
+			receivedSlotNumber := block.SlotNumber()
+
+			if receivedSlotNumber > c.lastSlot && receivedSlotNumber > 1 {
+				c.updateHead(receivedSlotNumber)
 			}
+			log.WithField("blockHash", fmt.Sprintf("0x%x", h)).Info("Received full block, processing validity conditions")
 
-			// TODO: Using random hash based on time stamp for seed, this will eventually be replaced by VDF or RNG.
-			// TODO: Uncomment after there is a reasonable way to bootstrap validators into the
-			// protocol. For the first few blocks after genesis, the current approach below
-			// will panic as there are no registered validators.
-			timestamp := time.Now().Unix()
-			activeState, err := c.chain.computeNewActiveState(common.BytesToHash([]byte(string(timestamp))))
+			// Process block as a validator if beacon node has registered, else process block as an observer.
+			if c.validator {
+				canProcess, err = c.chain.CanProcessBlock(c.web3Service.Client(), block, true)
```

Comment:
I don't think [this comment](https://github.com/prysmaticlabs/prysm/pull/422#discussion_r211939848) was addressed. Specifically, this validation step should come before `updateHead`.